utils/tfw_logger.cc

```diff
@@ -0,0 +1,463 @@
+	sigaction(SIGTERM, &sa, NULL);
+}
+
+void stop_daemon()
+{
+	pid_t pid;
+	std::ifstream pid_file(pid_file_path);
+
+	std::cout << "Stopping daemon..." << std::endl;
+
+	if (!pid_file)
+		throw Except("No PID file found. Is the daemon running?");
+
+	pid_file >> pid;
+	pid_file.close();
+
+	if (pid <= 0)
+		throw Except("Invalid PID in PID file.");
+
+	if (kill(pid, SIGTERM) < 0)
+		throw Except("Failed to stop daemon: {}", strerror(errno));
+
+	while (1) {
+		if (kill(pid, 0) == -1 && errno == ESRCH)
+			break;
+		std::this_thread::sleep_for(
+			std::chrono::milliseconds(WAIT_FOR_STOP));
+        }
+
+	std::cout << "Daemon stopped." << std::endl;
+}
+
+int
+main(int argc, char* argv[])
+{
+	std::vector<std::thread> thrs;
+	std::vector<std::future<void>> futures;
+	unsigned int i;
+	long int cpu_cnt;
+	int fd = -1, res = 0;
+	bool stop = false;
+	std::shared_ptr<spdlog::logger> logger = nullptr;
+
+	try {
+		po::options_description desc{"Usage: tfw_logger "
+					     "([options] <host> <log>) | --stop"};
+		desc.add_options()
+			("help,h", "show this message and exit")
+			("host,H", po::value<std::string>(),
+				"clickserver host address (required)")
+			("log,l", po::value<std::string>(),
+				"log path (required)")
+			("ncpu,n", po::value<unsigned int>(),
+				"manually specifying the number of CPUs")
+			("stop,s", po::bool_switch(&stop), "stop the daemon")
+			("user,u", po::value<std::string>(), "clickhouse user")
+			("password,p", po::value<std::string>(),
+			 "clickhouse password")
+			;
+		po::positional_options_description pos_desc;
+		pos_desc.add("host", 1);
+		pos_desc.add("log", -1);
+		po::variables_map vm;
+		po::store(po::command_line_parser(argc, argv)
+			.options(desc)
+			.positional(pos_desc)
+			.run(),
+			vm);
+		po::notify(vm);
+
+		if (stop) {
+			if (vm.size() > 1)
+				throw Except("--stop can't be used with "
+					     "another arguments");
+			stop_daemon();
+			return 0;
+		}
+
+		if (vm.count("help")) {
+			std::cout << desc << std::endl;
+			return 0;
+		}
+
+		if (!vm.count("host")) {
+			std::cerr << "'host' argument is requred" << std::endl;
+			return 1;
+		}
+
+		if (!vm.count("log"))
+			throw Except("please, specify log path");
+
+		if (vm.count("ncpu")) {
+			cpu_cnt = vm["ncpu"].as<unsigned int>();
+		} else {
+			cpu_cnt = sysconf(_SC_NPROCESSORS_ONLN);
+			if (cpu_cnt < 0)
+				throw Except("Can't get CPU number");
+		}
+
+		auto user = vm.count("user") ?
+			    vm["user"].as<std::string>() : std::string("");
+		auto password = vm.count("password") ?
+				vm["password"].as<std::string>() :
+				std::string("");
+
+		/*
+		 * When the daemon forks, it inherits the file descriptor for
+		 * /tmp/tempesta-lock-file, which was originally opened and locked by flock
+		 * in the tempesta.sh script. After daemonizing, the daemon process continues
+		 * to hold this lock, preventing subsequent executions of tempesta.sh.
+		 *
+		 * Close all descriptors before daemonizing.
+		 */
+		closefrom(3);
```

Comment:
It seems that we still have some problem with messaging. I try to start Tempesta FW with clickhouse, but I see that logs will be placed into the log file only after daemon stopped.

Reply:
Yes, I noticed it too.
`logger->flush_on(spdlog::level::info);` was needed.
Fixed.